test-harness/tests/test_cache.py

```diff
@@ -637,3 +638,83 @@ def test_if_failed_request_triggered_update_is_recovered_by_timers(
 
             # Verify that the cache is OK now
             ping_mesos_agent(ar, valid_user_header)
+
+    def test_if_early_boot_stage_is_not_sensitive_to_failures(
+            self, nginx_class, valid_user_header, mocker, log_catcher):
+        # The idea here is to make Backend so slow, that AR is unable to
+        # update cache on first request. AR should then switch to fail-fast
+        # behaviour and wait for timers to update the backend when it recovers.
+
+        refresh_lock_timeout = 10
+        backend_request_timeout = 5
+
+        ar = nginx_class(cache_first_poll_delay=1,
+                         cache_poll_period=3,
+                         cache_expiration=2,
+                         cache_max_age_soft_limit=1200,
+                         cache_max_age_hard_limit=1800,
+                         cache_backend_request_timeout=backend_request_timeout,
+                         cache_refresh_lock_timeout=refresh_lock_timeout,
+                         )
+        agent_id = 'de1baf83-c36c-4d23-9cb0-f89f596cd6ab-S1'
+        url = ar.make_url_from_path('/agent/{}/blah/blah'.format(agent_id))
+        v = Vegeta(log_catcher, target=url, jwt=valid_user_header, rate=3)
+
+        # Make mesos slooowwwww.....
+        # It mus respond slower than backend_request_timeout
+        mocker.send_command(endpoint_id='http://127.0.0.2:5050',
+                            func_name='always_stall',
+                            aux_data=backend_request_timeout * 1.5)
+
+        with GuardedSubprocess(ar):
+            with GuardedSubprocess(v):
+                time.sleep(1)  # let it warm-up!
+
+                # Vegeta by this time should have already clogged AR
```

Comment:
And if that is indeed what the test asserts, then you only need four requests in total:

The first one to trigger the cache update and block until it succeeds or fails.

The second and third to test that all queued requests fail immediately.

A fourth request which is expected to succeed N seconds later when the cache is expected to be populated (to test the case where mesos comes back after crashing.)